sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/TracerProvider.java

```diff
@@ -0,0 +1,175 @@
+                                                              String spanName,
+                                                              String databaseId,
+                                                              String endpoint) {
+        return traceEnabledPublisher(resultPublisher,  context, spanName, databaseId, endpoint, (T response) -> HttpConstants.StatusCodes.OK);
+    }
+
+    public <T> Mono<CosmosAsyncItemResponse<T>> traceEnabledCosmosItemResponsePublisher(Mono<CosmosAsyncItemResponse<T>> resultPublisher,
+                                                                                        Context context,
+                                                                                        String spanName,
+                                                                                        String databaseId,
+                                                                                        String endpoint) {
+        return traceEnabledPublisher(resultPublisher, context, spanName,databaseId, endpoint,
+            CosmosAsyncItemResponse::getStatusCode);
+    }
+
+    public <T> Mono<T> traceEnabledPublisher(Mono<T> resultPublisher,
+                                             Context context,
+                                             String spanName,
+                                             String databaseId,
+                                             String endpoint,
+                                             Function<T, Integer> statusCodeFunc) {
+        final AtomicReference<Context> parentContext = new AtomicReference<>(Context.NONE);
+        Optional<Object> callDepth = context.getData(COSMOS_CALL_DEPTH);
+        final boolean isNestedCall = callDepth.isPresent();
+        return resultPublisher
+            .doOnSubscribe(ignoredValue -> {
+                if (!isNestedCall) {
+                    parentContext.set(this.startSpan(spanName, databaseId, endpoint,
+                        context));
+                }
+            }).doOnSuccess(response -> {
+                if (!isNestedCall) {
+                    this.endSpan(parentContext.get(), Signal.complete(), statusCodeFunc.apply(response));
+                }
+            }).doOnError(throwable -> {
+                if (!isNestedCall) {
+                    this.endSpan(parentContext.get(), Signal.error(throwable), 0);
+                }
+            });
+    }
+
+    private void end(int statusCode, Throwable throwable, Context context) {
+        for (Tracer tracer : tracers) {
+            if (throwable != null) {
+                tracer.setAttribute(TracerProvider.ERROR_MSG, throwable.getMessage(), context);
+                tracer.setAttribute(TracerProvider.ERROR_TYPE, throwable.getClass().getName(), context);
+                StringWriter errorStack = new StringWriter();
+                throwable.printStackTrace(new PrintWriter(errorStack));
```

Comment:
do we think this is a good idea?
especially since `Throwable.printStackTrace()` isn't thread-safe and could give out confusing logs/statements?
cc: @srnagar

Reply:
I don't think we should have the entire stacktrace in tracing. It's important to capture that there was an error but the stracktrace is likely not that useful. Stacktrace should be available in logs, if that's necessary. It's also a security issue to capture stacktrace since the user cannot control or turn it off.

Reply:
Removed the error stacktrace